pkg/kclient/kclient.go

```diff
@@ -158,6 +161,12 @@ func NewForConfig(config clientcmd.ClientConfig) (client *Client, err error) {
 		return nil, err
 	}
 
+	config_flags := genericclioptions.NewConfigFlags(true)
+	client.cachedDiscoveryClient, err = config_flags.ToDiscoveryClient()
```

Comment:
Do we want to cache between calls to `odo`, or only during an odo execution.

If we want to cache only during an odo execution, it seems that we could call `Invalidate` just after creating the client

Reply:
> If we want to cache only during an odo execution, it seems that we could call `Invalidate` just after creating the client

If we were to invalidate the cache each time we create a client, I think the gain we see in `odo list` performance in this PR would be gone. But I need to check that to be sure.

Reply:
> I think the gain we see in `odo list` performance in this PR would be gone. But I need to check that to be sure.

12s without invalidating the cache, 16s if we invalidate it right after creation.
```sh
$ odo list
 ✓  Listing components from namespace 'myproject' [12s]
 NAME                     PROJECT TYPE  RUNNING IN  MANAGED 
 * devfile-nodejs-deploy  nodejs        Deploy      odo     

$ odo list
 ✓  Listing components from namespace 'myproject' [16s]
 NAME                     PROJECT TYPE  RUNNING IN  MANAGED 
 * devfile-nodejs-deploy  nodejs        Deploy      odo
```

Reply:
What is the client caching? the discovered resources (CRD level), or also the resources themselves (CR level)?

How is the cache consistent: Do we always get the latest versions of what we query (at the price of some back-and-forths to know if the cached one is the last one), of is there some period during which we get old versions?

Reply:
> What is the client caching? the discovered resources (CRD level), or also the resources themselves (CR level)?

You can look at the cache under `~/.kube/cache/discovery`. I took the cue from `kubectl api-resources cached=true` command.



> How is the cache consistent: Do we always get the latest versions of what we query (at the price of some back-and-forths to know if the cached one is the last one), of is there some period during which we get old versions?

There's a ttl, beyond which it falls back to using regular discovery client:
https://github.com/kubernetes/client-go/blob/master/discovery/cached/disk/cached_discovery.go#L155-L157

Reply:
> What is the TTL used?

![Screenshot from 2022-07-06 12-59-26](https://user-images.githubusercontent.com/713521/177496283-064a45ff-55c8-4152-9171-27017879a428.png)

I think it's the amount of nanoseconds you see in that screenshot.

> I think some documentation on this behaviour will be necessary, for the user to understand why `odo list` (and other commands) does not return the same values in the same circumstances

OK. But I don't follow you. Can you share an example of the `odo list` behaviour you're describing? As far as this PR is concerned, it's only caching the API resources. It's still making calls to the server to know if there are any instances of those resources. So, unless a user's adding an API resource within the TTL, I don't understand why odo would return differing values in any circumstances. However, as I said, I don't follow you entirely, so an example would be helpful.

Reply:
10 minutes, right? 
The output could differ after installing a new operator / CRD and deploying resources of this kind. These new resources should not be visible during 10 minutes

Reply:
Yes, and if possible, what are the solutions if the user doesn't want to wait for 10 minutes

Reply:
> > what are the solutions if the user doesn't want to wait for 10 minutes
> 
> We could document that the user could remove the cache directory from their system using `rm -rf` if they don't want to wait. However, this would be relevant for users using minikubs/CRC/kinD. Folks in the more realistic environment might not even have the permissions to install an Operator/CRD.
> 
> For the time being, I will add this info in `odo list` docs in this PR.

We discussed this topic when being in Paris, with Serena and the team. odo is dedicated to two different kind of people:

- the "lead" developer, who will experiment odo, create and update the Devfile, test different operators, bindings, etc. He will probably have admin access to the cluster to make experiments (on its own CRC cluster for example)
- the "normal" developer, who will have a tailor-made environment (prepared by the first "lead" developer"), and with minimal access to cluster.

The conclusion, as I remember, were that we need mostly to document for the "lead" developer, with the maximum of information for him to understand how to create and update the Devfile efficiently. The doc for the "normal" developer would be very simple, and would be prepared by the "lead" developer.

Reply:
Yes, I think so. Don't forget to say it is a cache common to the kubectl command

Reply:
OK.

Reply:
https://deploy-preview-5908--odo-docusaurus-preview.netlify.app/docs/command-reference/list